state/status.go

```diff
@@ -199,48 +199,75 @@ type statusHistoryArgs struct {
-	filter := args.filter
-	if err := args.filter.Validate(); err != nil {
-		return nil, errors.Annotate(err, "validating arguments")
-	}
-	statusHistory, closer := args.st.getCollection(statusesHistoryC)
-	defer closer()
-
+func fetchNStatusResults(size int, key string, timeDelta *time.Duration,
+	date *time.Time, col mongo.Collection) ([]historicalStatusDoc, error) {
 	var (
 		docs  []historicalStatusDoc
 		query mongo.Query
 	)
-	baseQuery := bson.M{"globalkey": args.globalKey}
-	if filter.Delta != nil {
-		delta := *filter.Delta
-		// TODO(perrito666) 2016-05-02 lp:1558657
+	baseQuery := bson.M{"globalkey": key}
+	if timeDelta != nil {
+		delta := *timeDelta
+		// TODO(perrito666) 2016-10-06 lp:1558657
 		updated := time.Now().Add(-delta)
-		baseQuery = bson.M{"updated": bson.M{"$gt": updated.UnixNano()}, "globalkey": args.globalKey}
+		baseQuery = bson.M{"updated": bson.M{"$gt": updated.UnixNano()}, "globalkey": key}
 	}
-	if filter.Date != nil {
-		baseQuery = bson.M{"updated": bson.M{"$gt": filter.Date.UnixNano()}, "globalkey": args.globalKey}
+	if date != nil {
+		baseQuery = bson.M{"updated": bson.M{"$gt": date.UnixNano()}, "globalkey": key}
 	}
-	query = statusHistory.Find(baseQuery).Sort("-updated")
-	if filter.Size > 0 {
-		query = query.Limit(filter.Size)
+	query = col.Find(baseQuery).Sort("-updated")
+	if size > 0 {
+		query = query.Limit(size)
 	}
 	err := query.All(&docs)
 
 	if err == mgo.ErrNotFound {
-		return []status.StatusInfo{}, errors.NotFoundf("status history")
+		return []historicalStatusDoc{}, errors.NotFoundf("status history")
 	} else if err != nil {
-		return []status.StatusInfo{}, errors.Annotatef(err, "cannot get status history")
+		return []historicalStatusDoc{}, errors.Annotatef(err, "cannot get status history")
 	}
+	return docs, nil
+
+}
 
-	results := make([]status.StatusInfo, len(docs))
-	for i, doc := range docs {
-		results[i] = status.StatusInfo{
-			Status:  doc.Status,
-			Message: doc.StatusInfo,
-			Data:    utils.UnescapeKeys(doc.StatusData),
-			Since:   unixNanoToTime(doc.Updated),
+func statusHistory(args *statusHistoryArgs) ([]status.StatusInfo, error) {
+	filter := args.filter
+	if err := args.filter.Validate(); err != nil {
+		return nil, errors.Annotate(err, "validating arguments")
+	}
+	statusHistory, closer := args.st.getCollection(statusesHistoryC)
+	defer closer()
+
+	var results []status.StatusInfo
+	// Just attempt to get the desired amount of results a couple of times
+	// to avoid dead locks.
+	for i := 0; i < 3; i++ {
+		size := args.filter.Size
+		if len(results) > 0 {
+			size = size - len(results)
+		}
+		if size < 1 && args.filter.Size != 0 {
+			break
+		}
+
+		docs, err := fetchNStatusResults(size, args.globalKey,
+			args.filter.Delta, args.filter.Date, statusHistory)
+		partial := []status.StatusInfo{}
+		if err != nil {
+			return []status.StatusInfo{}, errors.Trace(err)
+		}
+		for _, doc := range docs {
+			if filter.Exclude.Contains(doc.StatusInfo) {
```

Comment:
Why do this here? Can't it be made part of the query expression? The query would then be atomic.